File: hashmap.c
#include <stdio.h>
#include <stdlib.h>
// #include "hashmap.h"
// this header is auto generated
#include "pointer/hashmap.h"

int main(int argc, char* argv[]) {
    printf("C: Hello, world!\n");
    const void* map = build_hashmap();
    printf("C: hashmap: %p\n", map);
    // ask for a key while q is pressed
    char buffer[256];
    while (1) {
        printf("C: key: ");
        // this should be a number, ignore other characters unless it is q
        fgets(buffer, 256, stdin);
        if (buffer[0] == 'q') {
            break;
        }

        int key = atoi(buffer);
        const Foo* foo = get_hashmap_value(map, key);

        if (foo == NULL) {
            printf("C: value is null\n");
        } else {
            printf("C: bytes: [ ");
            // print the next 12 bytes
            for (int i = 0; i < 12; i++) {
                printf("%d ", ((unsigned char*)foo)[i]);
            }
            printf("]\n");
            printf("C: value pointer: %p\n", foo);
            printf("C: Foo: { x: %d, y: Bar { x: %d, y: %d } }\n", foo->x, foo->y.x, foo->y.y);
            free_foo(foo);
            printf("C: value freed\n");
        }
    }
    free_hashmap(map);
    printf("C: hashmap freed\n");
    return 0;
}

File: hashmap.dart
import 'dart:ffi';
import 'dart:io';

import 'hashmap_bindings.dart';

void main(List<String> arguments) {
  print("Dart: Hello World");
  final hashmap = RustHashmap(DynamicLibrary.open("pointer.dll"));
  final pointer = hashmap.build_hashmap();
  print("Dart: Pointer: ${pointer.address.toRadixString(16)}");
  while (true) {
    stdout.write("Dart: Enter a key: ");
    final key = stdin.readLineSync();
    if (key == null) break;
    if (key.startsWith('q')) {
      break;
    }

    final value = int.tryParse(key);
    if (value == null) {
      print("Dart: Invalid key");
      continue;
    }

    final result = hashmap.get_hashmap_value(pointer, value);
    if (result.address == 0) {
      print("Dart: Key not found");
    } else {
      stdout.write("Dart: bytes: [ ");
      final bytes = result.cast<Uint8>();
      for (var i = 0; i < 12; i++) {
        final byte = bytes.elementAt(i).value;
        stdout.write("$byte ");
      }
      print("]");
      print("Dart: value pointer: ${result.address.toRadixString(16)}");
      // cast to Foo
      final foo = result.cast<Foo>();
      print(
        "Dart: Foo: { x: ${foo.ref.x}, y: Bar: { x: ${foo.ref.y.x}, y: ${foo.ref.y.y} } }",
      );
      hashmap.free_foo(result);
      print("Dart: value freed");
    }
  }
  hashmap.free_hashmap(pointer);
  print("Dart: hashmap freed");
}
